File: 567-permutation-in-string/permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        
        s1_count = Counter(s1)
        s2_count = defaultdict(int)
        L = 0

        for R in range(len(s2)):
            c = s2[R]
            s2_count[c] += 1

            if (R - L + 1) > len(s1):
                c = s2[L]
                s2_count[c] -= 1
                if not s2_count[c]:
                    del s2_count[c]
                L += 1
            
            if s1_count == s2_count: return True
        return False

File: 567-permutation-in-string/permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s2.size() < s1.size()) {
            return false;
        }

        unordered_map<char, int> s1_cnt;
        for (char c : s1) {
            s1_cnt[c]++;
        }

        int L = 0;
        unordered_map<char, int> s2_cnt;
        for (int R = 0; R < s2.size(); ++R) {
            s2_cnt[s2[R]]++;
            if ((R - L + 1) > s1.size()) {
                s2_cnt[s2[L]]--;
                if (!s2_cnt[s2[L]]) {
                    s2_cnt.erase(s2[L]);
                }
                L++;
            }

            if (s2_cnt == s1_cnt) {
                return true;
            }
        }

        return false;
    }
};